### ISSUE TITLE
Is possible to use Numba to speed up FinRL?

### ISSUE THREAD

**Issue by dev590t** [AI4Finance-Foundation/FinRL#613]
The preprocessing is very slow in FinRL. The Nmba allow GPU parallelization with C speed for process array data. Some high speed trading framework like vectorbt use it. 

Is possible to take experience from vectorbt and use Numba to speed up FinRL?

**Comment by XiaoYangLiu-FinRL** [comment]
https://github.com/AI4Finance-Foundation/ElegantRL
If you check ElegantRL, you will find more advanced way of using GPU, DGX-2 server and even NVIDIA's SuperPod. Enjoy!

**Comment by dev590t** [comment]
If I understood the code correctly,ElegantRL is used in DRL step. I mean use Numaba in preprocessing data step, like here:
```
data = DP.clean_data(data)
data = DP.add_technical_indicator(data, INDICATORS)
data = DP.add_vix(data)
```
Did this part of code use also GPU acceleration?

**Comment by XiaoYangLiu-FinRL** [comment]
Yes, ElegantRL is for DRL algorithms.  Your suggestions for accelerating data preprocessing are quite interesting.  Looking forward to workable codes, would that be possible from your end?

**Comment by zhumingpassional** [comment]
> @XiaoYangLiu-FinRL If I understood the code correctly,ElegantRL is used in DRL step. I mean use Numba in preprocessing data step, like here:
> 
> ```
> data = DP.clean_data(data)
> data = DP.add_technical_indicator(data, INDICATORS)
> data = DP.add_vix(data)
> ```
> 
> Did this part of code use also GPU acceleration?

@dev590t 
Absolutely not. The current does not support numba. We may test the acceleration to decide whether to add numba.

**Comment by dev590t** [comment]
Unfortunaly, I'm not a python data scientist programmer, and have never used Numba. I have discover this lib from the trading lib vectorbt.

vectorbt can also preprocess data. I think it is maybe possible to use the dataframe produced by vectorbt and pass it directly to DRL agent,  instead of use dataframe produced by the built-in code of FinRL.

This have few advandage:
- GPU acceleration in preprocessing
- it maybe support more indicator than FinRL.  According its doc, it support for 99% indicators in Technical Analysis Library, Pandas TA, and TA-Lib. So it is possible to pass more indicators to DRL agent.

Do you think it is a good direction to explore?

**Comment by zhumingpassional** [comment]
Yes. it is a good direction. Thanks for your suggestions.

**Comment by zhumingpassional** [comment]
Some engineers said that they tried numba, however there exist errors if the scenario becomes complex. They recommend cupy and cython.

**Comment by dev590t** [comment]
I think it is better to test in first existing GPU trading indicator librairy before to decide to reimplement or no the preprocessing step with Numba or Cupy. 
And if existing GPU trading indicator librairy is not suitable, Cupy also could be a interested solution.

**Comment by zhumingpassional** [comment]
With respect to existing GPU trading indicator librairy, could you give any examples? What about the performance?
